python/ray/data/_internal/planner/exchange/push_based_shuffle_task_scheduler.py

```diff
@@ -621,39 +706,65 @@ def _compute_shuffle_schedule(
         num_output_blocks: int,
     ) -> _PushBasedShuffleStage:
         num_cpus_total = sum(v for v in num_cpus_per_node_map.values())
+        logger.get_logger().info(
+            f"Found {num_cpus_total} CPUs available CPUs for push-based shuffle."
+        )
         task_parallelism = min(num_cpus_total, num_input_blocks)
 
         num_tasks_per_map_merge_group = merge_factor + 1
+        num_total_merge_tasks = math.ceil(num_input_blocks / merge_factor)
+
         num_merge_tasks_per_round = 0
         merge_task_placement = []
         leftover_cpus = 0
         # Compute the total number of merge tasks and their node placement.
         # Each merge task should be grouped with `merge_factor` map tasks for
         # pipelining. These groups should then be spread across nodes according
         # to CPU availability for load-balancing.
+        num_input_blocks_remaining = num_input_blocks
         for node, num_cpus in num_cpus_per_node_map.items():
-            node_parallelism = min(
-                num_cpus, num_input_blocks // len(num_cpus_per_node_map)
+            # First find how many merge tasks we should run on this node.
+            # We take the min of the number of CPUs on this node and the number
+            # of input blocks that we haven't scheduled yet, in case there are
+            # fewer input blocks than CPU slots on this node.
+            num_cpu_slots = min(num_cpus, num_input_blocks_remaining)
+            num_merge_tasks_on_cur_node = int(
+                num_cpu_slots // num_tasks_per_map_merge_group
             )
-            num_merge_tasks = node_parallelism // num_tasks_per_map_merge_group
-            for i in range(num_merge_tasks):
+            # For small datasets, the number of tasks to run may be less than
+            # the total CPU slots available.
+            num_merge_tasks_on_cur_node = min(
+                num_merge_tasks_on_cur_node, num_total_merge_tasks
+            )
+            for i in range(num_merge_tasks_on_cur_node):
                 merge_task_placement.append(node)
-            num_merge_tasks_per_round += num_merge_tasks
+                num_input_blocks_remaining -= merge_factor
```

Comment:
Could you explain this line means? and is this still correct when merge_factor is a float?

Reply:
Added comment, a safety check to round up num_input_blocks_remaining and tests.